User guide updates membrane (#1933)
OpenFreeEnergy/openfe: docs/guide/setup/chemical_systems_and_thermodynamic_cycles.rst

:class:`.Component`\s, which describe what is physically present in a system;
:class:`.ChemicalSystem`\s, which combine components into a complete end state;
and thermodynamic cycles, which connect end states via alchemical transformations.

.. _userguide_components:

Components
----------

Components are the composable building blocks that define the chemical
composition of a simulated system. Splitting a system into components serves three purposes:

1. Alchemical transformations can be easily understood by comparing the differences in components.
2. Components can be reused to compose different systems.
3. :class:`.Protocol`\s can apply component-specific behaviour, e.g. different force fields per component.

Component types — overview
~~~~~~~~~~~~~~~~~~~~~~~~~~

.. list-table::
   :header-rows: 1
   :widths: 25 30 45

   * - Component
     - Role
     - Key notes
   * - :class:`.ProteinComponent`
     - Biological assembly
     - Typically the contents of a PDB file. May include crystallographic waters and ions (defined as HETATM entries),
       and disulfide bonds (defined via CONECT records).
   * - :class:`.SmallMoleculeComponent`
     - Ligands and cofactors

>      -

Reviewer: Do we want to say that they can optionally contain atomic partial charges which can be used in the simulation?

Author: Great idea, added this!

     membrane_protein = openfe.ProteinMembraneComponent.from_pdb_file('./protein_membrane.pdb')

2. **Manual specification** — box vectors can be provided explicitly as numpy arrays with OpenFF units in OpenMM format via the ``box_vectors`` argument::

     import numpy as np
     import openff.units as offunit

     box_vectors = np.array([
         [6.9587, 0.0, 0.0],
         [0.0, 5.9164, 0.0],
         [0.0, 0.0, 9.2692]
     ]) * offunit.nanometer

     membrane_protein = openfe.ProteinMembraneComponent.from_pdb_file(
         './protein_membrane.pdb', box_vectors=box_vectors
     )

3. **Inference from atomic coordinates** — box vectors can be estimated from the atomic
   positions by passing ``infer_box_vectors=True``::

     membrane_protein = openfe.ProteinMembraneComponent.from_pdb_file(
         './protein_membrane.pdb', infer_box_vectors=True
     )

   .. warning::

      Inferring box vectors from atomic positions can be inaccurate if the PDB originates
      from a previous simulation where atoms may be distributed across periodic images.

.. _userguide_chemical_systems:

ChemicalSystem
--------------

> A :class:`.ChemicalSystem` is composed of components that together describe the full

Reviewer: ```suggestion
A :class:`.ChemicalSystem` is composed of components that together describe a model of the system to be simulated.
```

Reviewer: Maybe spliting this into model building and simulation parameters might be a good way to get the message accross?

simulated system. It represents the **end state** of an alchemical transformation
and is the primary input a :class:`.Protocol` consumes to define a simulation state.

**What a ChemicalSystem defines**

* Exact atomic information (including protonation state) of protein, ligands,
  cofactors, and any crystallographic waters.
* Atomic positions of all explicitly defined components such as ligands or proteins.
* The abstract or explicit definition of the solvent environment (SolventComponent).

**What a ChemicalSystem does NOT define**, and are instead handled by the Protocol:

> * Forcefield applied to any component, including water model or virtual particles.

Reviewer: ```suggestion
Any simulation parameters including:
* Forcefield applied to any component, including water model or virtual particles.
```

.. _userguide_chemicalsystems_and_components:

Components, Chemical Systems and Thermodynamic Cycles
=====================================================

This page describes the core building blocks used to define simulation states in openfe:
:class:`.Component`\s, which describe what is physically present in a system;
:class:`.ChemicalSystem`\s, which combine components into a complete end state;
and thermodynamic cycles, which connect end states via alchemical transformations.

.. _userguide_components:

Components
----------

> Components are the additive, composable building blocks that define the chemical

Reviewer 2: Might be ok to just say composable and remove additive - not sure if there are synonyms.

composition of a simulated system. Splitting a system into components serves three purposes:

1. Alchemical transformations can be easily understood by comparing the differences in components.
2. Components can be reused to compose different systems.
3. :class:`.Protocol`\s can apply component-specific behaviour, e.g. different force fields per component.

Component types — overview
~~~~~~~~~~~~~~~~~~~~~~~~~~

.. list-table::
   :header-rows: 1
   :widths: 25 30 45

   * - Component
     - Role
     - Key notes
   * - :class:`.ProteinComponent`
     - Biological assembly

>      - Typically the contents of a PDB file. May include crystallographic waters, ions,

Reviewer 2: maybe add "(defined as HETATM entries)" after the waters and ions?

  during system preparation.
* **Explicit solvation** — use a :class:`.SolvatedPDBComponent` or
  :class:`.ProteinMembraneComponent`. Solvent molecule coordinates (waters and ions) are explicitly defined in the inputs.

Either define the solvent abstractly, or provide a fully solvated system — do not mix
both for the same leg of a transformation.

.. note::

   Some protocols, such as :class:`.SepTopProtocol` and :class:`.AbsoluteBindingProtocol`,
   use a single :class:`.ChemicalSystem` to represent both the complex and solvent legs.
   In this case, a :class:`.ChemicalSystem` may contain both a :class:`.SolventComponent`
   and a :class:`.ProteinMembraneComponent`. However, these apply to *different* legs: the
   :class:`.SolventComponent` is used only for the solvent leg, and the
   :class:`.ProteinMembraneComponent` (which is already explicitly solvated) is used only
   for the complex leg. The mutual exclusivity rule still holds per leg.

Box vectors for explicitly solvated systems
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The components :class:`.SolvatedPDBComponent` and :class:`.ProteinMembraneComponent`

> requires periodic box vectors. These can be provided in three ways:

Reviewer 3: ```suggestion
require periodic box vectors. These can be provided in three ways:
```

1. **CRYST record in the PDB file** — OpenMM reads box vectors automatically. No additional arguments are needed::

     membrane_protein = openfe.ProteinMembraneComponent.from_pdb_file('./protein_membrane.pdb')

> 2. **Manual specification** — box vectors can be provided explicitly in OpenMM format  via the ``box_vectors`` argument::

Reviewer 2: mention box being defined with OpenFF units somewhere here?

         './protein_membrane.pdb', box_vectors=box_vectors
     )

3. **Inference from atomic coordinates** — box vectors can be estimated from the atomic
   positions by passing ``infer_box_vectors=True``::

     membrane_protein = openfe.ProteinMembraneComponent.from_pdb_file(
         './protein_membrane.pdb', infer_box_vectors=True
     )

   .. warning::

      Inferring box vectors from atomic positions can be inaccurate if the PDB originates
      from a previous simulation where atoms may be distributed across periodic images.

.. _userguide_chemical_systems:

ChemicalSystem
--------------

A :class:`.ChemicalSystem` is composed of components that together describe the full
simulated system. It represents the **end state** of an alchemical transformation
and is the primary input a :class:`.Protocol` consumes to define a simulation state.

**What a ChemicalSystem defines**

* Exact atomic information (including protonation state) of protein, ligands,
  cofactors, and any crystallographic waters.
* Atomic positions of all explicitly defined components such as ligands or proteins.
* The abstract or explicit definition of the solvent environment (SolventComponent).

> **What a ChemicalSystem does NOT define**

Reviewer 3: ```suggestion
**What a ChemicalSystem does NOT define**, and are instead handled by the Protocol:
```

* Forcefield applied to any component, including water model or virtual particles.
* Thermodynamic conditions (e.g. temperature or pressure).

> * These are handled by the :class:`.Protocol`.

Reviewer 3: ```suggestion
* These are handled by the :class:`.Protocol`.
```

nitpick - I think it's easier to understand if this detail is at the top. Please ignore if you disagree!

   * - **Standard protein–ligand**
     - | **Complex leg:**
       | :class:`.ProteinComponent` + :class:`.SmallMoleculeComponent`\s + :class:`.SolventComponent`
       |
       | **Solvent leg:**
       | :class:`.SmallMoleculeComponent`\s + :class:`.SolventComponent`
     - | **Single ChemicalSystem (both legs):**
       | :class:`.ProteinComponent` + :class:`.SmallMoleculeComponent`\s + :class:`.SolventComponent`
   * - **Membrane system**
     - | **Complex leg:**
       | :class:`.ProteinMembraneComponent` + :class:`.SmallMoleculeComponent`\s
       | *(no* :class:`.SolventComponent` *— already explicitly solvated)*
       |
       | **Solvent leg:**
       | :class:`.SmallMoleculeComponent`\s + :class:`.SolventComponent`
     - | **Single ChemicalSystem (both legs):**
       | :class:`.ProteinMembraneComponent` + :class:`.SmallMoleculeComponent`\s + :class:`.SolventComponent`
       | *(protocol applies* :class:`.SolventComponent` *only in the solvent leg)*

Thermodynamic Cycles
--------------------

A thermodynamic cycle can be described as a set of :class:`.ChemicalSystem`\s (nodes) connected by
alchemical transformations (edges). The :class:`.Protocol` defines how the

> :class:`.ChemicalSystem`\s map onto the cycle and how they are used in practice.

Reviewer 3: I think this unclear regarding thermodynamic cycles and alchemical networks. Not blocking for this PR, but maybe we can follow up with a better explainer over in the alchemical network docs?

Author: I ended up removing the reference to the alchemical networks here, because I think it would confuse users more, since a thermodynamic cycle is more linked to a Transformation. I think the reference to the protocols here is more helpful, since e.g. in the RBFE protocol we say how the ChemicalSystems are used to describe the thermodynamic cycle.

See :ref:`the user guide for alchemical networks <alchemical_network_model>` for more information.

The same :class:`.ChemicalSystem` can be reused across multiple thermodynamic states
depending on the protocol. For details of which end states to construct, consult the
:ref:`pages for each specific Protocol <userguide_protocols>`.

> RBFE example

Reviewer 2: hybrid topology RBFE?

* forcefield applied to any component, including details on water model or virtual particles
* thermodynamic conditions (e.g. temperature or pressure)

For some protocols, such as :class:`.SepTopProtocol` and :class:`.AbsoluteBindingProtocol`, a single
:class:`.ChemicalSystem` is used to represent both legs of the thermodynamic cycle (complex and solvent).
This is in contrast to the :class:`.RelativeHybridTopologyProtocol`, where each leg is defined by

> separate :class:`.ChemicalSystem`\s.

Reviewer 2: Maybe put this in a `.. note:` section? It would be good to say that this is subject to change too.

Author: Done!

.. _userguide_components:

Components
----------

A :class:`.ChemicalSystem` is composed of many 'component' objects, which together define the overall system.

Examples of components include:

> * :class:`.ProteinComponent`: an entire biological assembly, typically the contents of a PDB file.

Reviewer: we could hint here that the ProteinComponent can also contain crystal waters.

Author: Added this!

* :class:`.SmallMoleculeComponent`: typically ligands and cofactors
* :class:`.SolventComponent`: solvent conditions
* :class:`.ProteinMembraneComponent`: an explicitly solvated protein-membrane system, including box vectors.

Splitting the total system into components serves three purposes:

1. alchemical transformations can be easily understood by comparing the differences in components.
2. components can be reused to compose different systems.
3. :class:`.Protocol`\s can have component-specific behavior. E.g. different force fields for each component.

> When dealing with membrane-protein systems, the protein is represented using an explicitly solvated

Reviewer 2: The content of what is being added looks good, but there's a "missing link".

What I think we've been missing (since before membrane support) is the idea that the Components are composable. What would you think about something like this:

- These Components are additively added into a ChemicalSystem, which will define the chemical composition of the simulated system.
- For a conventional protein-ligand complex in water, this would look like [add protein, small and solvent components]
  - Some mention that the protein component handles crystallographic waters and also can define SSBONDs via the CONECT records.
- For membrane proteins: what you have here

:class:`.ProteinMembraneComponent` instead of a :class:`.ProteinComponent`.

The :class:`.ProteinMembraneComponent` requires periodic box vectors to define the simulation box.
There are several ways to provide these vectors:

1. CRYST record in the PDB file

   If the PDB file includes a CRYST record, OpenMM can automatically read the box vectors from it.

2. Manually specifying box vectors

   Box vectors can be provided explicitly in OpenMM format.

3. Inferring from atomic positions

   Box vectors can be estimated from the atomic coordinates in the PDB file.

>    Caveat: This approach can be inaccurate if the PDB comes from a previous simulation and some atoms are positioned in periodic images.

Reviewer: I would use a warning box for this to make it clear this should be a last resort.

Author: Added a warning box.

  ligand_B_complex = openfe.ChemicalSystem(components={'ligand': ligand_B, 'protein': protein, 'solvent': solvent})
  # ligand_A + solvent
  ligand_A_solvent = openfe.ChemicalSystem(components={'ligand': ligand_A, 'solvent': solvent})
  # ligand_B + solvent
  ligand_B_solvent = openfe.ChemicalSystem(components={'ligand': ligand_B, 'solvent': solvent})

> For a protein-membrane complex, the `ligand_A_complex` would instead use a

Reviewer 2: Thinking about it a bit more - will the code changes work for just any ExplicitPDBComponent? I.e. could this be reduced to just "if you explicitly defined the system components, e.g. with a protein membrane" and then "here is an example using a ProteinMembraneComponent".

Author: Changed this!
